cmd/snap-update-ns/bootstrap.c

```diff
@@ -75,6 +77,72 @@ setns_into_snap(const char* snap_name)
     return err;
 }
 
+// switch_to_privileged_user drops to the real user ID while retaining
+// CAP_SYS_ADMIN, to operations such as mount().
+static int
+switch_to_privileged_user()
+{
+    uid_t real_uid;
+    gid_t real_gid;
+
+    real_uid = getuid();
+    if (real_uid == 0) {
+        // We're running as root: no need to switch IDs
+        return 0;
+    }
+    real_gid = getgid();
+
+    // _LINUX_CAPABILITY_VERSION_3 valid for kernel >= 2.6.26. See
+    // https://github.com/torvalds/linux/blob/master/kernel/capability.c
+    struct __user_cap_header_struct hdr = { _LINUX_CAPABILITY_VERSION_3, 0 };
+    struct __user_cap_data_struct data[2] = { { 0 } };
+
+    data[0].effective = (CAP_TO_MASK(CAP_SYS_ADMIN) |
+                         CAP_TO_MASK(CAP_SETUID) |
+                         CAP_TO_MASK(CAP_SETGID));
+    data[0].permitted = data[0].effective;
+    data[0].inheritable = 0;
+    data[1].effective = 0;
+    data[1].permitted = 0;
+    data[1].inheritable = 0;
+
+    if (capset(&hdr, data) != 0) {
+        bootstrap_errno = errno;
+        bootstrap_msg = "cannot set permitted capabilities mask";
+        return -1;
+    }
+
+    if (setgroups(0, NULL) != 0) {
```

Comment:
This is only allowed on Linux (for portability would use `setgroups(1, gid_list)` where gid_list is a one element array of gid_t with the gid you want to drop to (ie, so getgroups() has a 1 element list with the euid)), but snapd is Linux-only, so +1. Perhaps add a comment to this effect?